In [2]:
#pip install torch torchvision torchaudio

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 59.6 MB 36.9 MB/s eta 0:00:01
     |████████████████████████████████| 1.5 MB 45.4 MB/s eta 0:00:01
     |████████████████████████████████| 1.7 MB 3.1 MB/s eta 0:00:01
  Using cached fsspec-2023.12.0-py3-none-any.whl (168 kB)
  Using cached filelock-3.13.1-py3-none-any.whl (11 kB)
     |████████████████████████████████| 133 kB 3.1 MB/s eta 0:00:01
     |████████████████████████████████| 1.6 MB 44.3 MB/s eta 0:00:01
     |████████████████████████████████| 5.7 MB 49.5 MB/s eta 0:00:01
     |████████████████████████████████| 3.3 MB 57.7 MB/s eta 0:00:01
     |████████████████████████████████| 62 kB 6.9 MB/s  eta 0:00:01
     |████████████████████████████████| 14.0 MB 57.7 MB/s eta 0:00:01
     |████████████████████████████████| 120 kB 33.4 MB/s eta 0:00:01
     |████████████████████████████████| 104 kB 36.7 MB/s eta 0:00:01
     |████████████████████████████████| 61 kB 625 

In [3]:
#pip install transformers

Defaulting to user installation because normal site-packages is not writeable
  Using cached transformers-4.35.2-py3-none-any.whl (7.9 MB)
     |████████████████████████████████| 426 kB 4.0 MB/s eta 0:00:01
     |████████████████████████████████| 78 kB 12.6 MB/s eta 0:00:01
  Using cached huggingface_hub-0.19.4-py3-none-any.whl (311 kB)
     |████████████████████████████████| 291 kB 40.8 MB/s eta 0:00:01
     |████████████████████████████████| 174 kB 37.5 MB/s eta 0:00:01
     |████████████████████████████████| 2.5 MB 40.7 MB/s eta 0:00:01
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [4]:
#pip install --upgrade transformers

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [5]:
#pip show transformers

Name: transformers
Version: 4.35.2
Summary: State-of-the-art Machine Learning for JAX, PyTorch and TensorFlow
Home-page: https://github.com/huggingface/transformers
Author: The Hugging Face team (past and future) with the help of all our contributors (https://github.com/huggingface/transformers/graphs/contributors)
Author-email: transformers@huggingface.co
License: Apache 2.0 License
Location: /Users/kariprimiano/Library/Python/3.9/lib/python/site-packages
Requires: tqdm, regex, tokenizers, pyyaml, numpy, safetensors, packaging, filelock, huggingface-hub, requests
Required-by: 
Note: you may need to restart the kernel to use updated packages.


In [6]:
#pip install --upgrade s3fs

Defaulting to user installation because normal site-packages is not writeable
  Using cached s3fs-2023.12.0-py3-none-any.whl (29 kB)
  Using cached aiobotocore-2.8.0-py3-none-any.whl (75 kB)
     |████████████████████████████████| 387 kB 4.9 MB/s eta 0:00:01
  Using cached botocore-1.33.1-py3-none-any.whl (11.6 MB)
     |████████████████████████████████| 46 kB 13.4 MB/s eta 0:00:01
     |████████████████████████████████| 81 kB 26.0 MB/s eta 0:00:01
     |████████████████████████████████| 61 kB 17.1 MB/s eta 0:00:01
     |████████████████████████████████| 143 kB 38.2 MB/s eta 0:00:01
  Attempting uninstall: urllib3
    Found existing installation: urllib3 2.1.0
    Uninstalling urllib3-2.1.0:
      Successfully uninstalled urllib3-2.1.0
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [7]:
#pip install fsspec==2023.3.0

Defaulting to user installation because normal site-packages is not writeable
  Using cached fsspec-2023.3.0-py3-none-any.whl (145 kB)
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2023.12.0
    Uninstalling fsspec-2023.12.0:
      Successfully uninstalled fsspec-2023.12.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
s3fs 2023.12.0 requires fsspec==2023.12.0, but you have fsspec 2023.3.0 which is incompatible.
huggingface-hub 0.19.4 requires fsspec>=2023.5.0, but you have fsspec 2023.3.0 which is incompatible.
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [8]:
#pip install nltk

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 1.5 MB 4.5 MB/s eta 0:00:01
  Using cached click-8.1.7-py3-none-any.whl (97 kB)
     |████████████████████████████████| 302 kB 44.8 MB/s eta 0:00:01
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [10]:
#pip install pandas

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 11.0 MB 4.1 MB/s eta 0:00:01
     |████████████████████████████████| 341 kB 40.4 MB/s eta 0:00:01
     |████████████████████████████████| 502 kB 52.2 MB/s eta 0:00:01
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [11]:
#pip install numpy

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [13]:
#pip install scikit-learn

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [1]:
from transformers import pipeline
import nltk
from nltk.corpus import stopwords
import string
import requests
import pandas as pd
import time
import json
import zipfile
import os
import re
import ast
import torch


import warnings
warnings.filterwarnings('ignore')

print(torch.__version__)


/Users/kariprimiano/Library/Python/3.9/lib/python/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


2.1.1


In [4]:
#master_data = pd.read_csv('master_anime_data.csv')

OSError: [Errno 89] Operation canceled

In [ ]:
# Sample a subset of your data for processing
#sampled_data = master_data.sample(frac=0.1, random_state=13)

In [2]:
sampled_data = pd.read_csv('sample_dataset.csv')

In [3]:
# Download nltk stop words
nltk.download('stopwords')

# Load the emotion classification pipeline
model_name = "j-hartmann/emotion-english-distilroberta-base"
emotion_classifier = pipeline('text-classification', model=model_name, return_all_scores=True)

# Set English stop words
stop_words = set(stopwords.words('english'))

# Function for advanced text cleaning
def clean_text(text):
    text = text.lower()
    text = re.sub(r'<[^>]+>', '', text)
    text = text.translate(str.maketrans('', '', string.punctuation))
    text = ' '.join([word for word in text.split() if word not in stop_words])
    return text

# Function to classify emotions of a text
def classify_emotions(text):
    cleaned_text = clean_text(text)  # Clean the text using clean_text function
    predictions = emotion_classifier(cleaned_text)
    return {emotion['label']: emotion['score'] for emotion in predictions[0]}

# Apply the model to the 'synopsis' column
sampled_data['emotion_scores'] = sampled_data['synopsis'].astype(str).apply(classify_emotions)

# Inspect results
sampled_data[['synopsis', 'emotion_scores']].head()


[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/kariprimiano/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


synopsis  \
0  After the accident in which she lost her mothe...   
1  Naota Nandaba is an ordinary sixth grader livi...   
2  Story begins as Momiji's chastity is taken by ...   
3  Tanya Degurechaff is a young soldier infamous ...   
4  Hachiman Hikigaya is an apathetic high school ...   

                                      emotion_scores  
0  {'anger': 0.010413373820483685, 'disgust': 0.0...  
1  {'anger': 0.18671022355556488, 'disgust': 0.00...  
2  {'anger': 0.007550527341663837, 'disgust': 0.0...  
3  {'anger': 0.5558895468711853, 'disgust': 0.003...  
4  {'anger': 0.011294235475361347, 'disgust': 0.0...

In [4]:
sampled_data['emotion_scores'].value_counts

<bound method IndexOpsMixin.value_counts of 0        {'anger': 0.010413373820483685, 'disgust': 0.0...
1        {'anger': 0.18671022355556488, 'disgust': 0.00...
2        {'anger': 0.007550527341663837, 'disgust': 0.0...
3        {'anger': 0.5558895468711853, 'disgust': 0.003...
4        {'anger': 0.011294235475361347, 'disgust': 0.0...
                               ...                        
34983    {'anger': 0.0200864989310503, 'disgust': 0.000...
34984    {'anger': 0.007323819678276777, 'disgust': 0.0...
34985    {'anger': 0.005832878407090902, 'disgust': 0.0...
34986    {'anger': 0.004111292771995068, 'disgust': 0.0...
34987    {'anger': 0.022216347977519035, 'disgust': 0.0...
Name: emotion_scores, Length: 34988, dtype: object>

### Expanding 'emotion_scores' into separate columns

In [5]:
# Check if the emotion_scores column needs to be parsed from string to dictionary
if isinstance(sampled_data['emotion_scores'].iloc[0], str):
    sampled_data['emotion_scores'] = sampled_data['emotion_scores'].apply(ast.literal_eval)

# Expand the emotion_scores column into separate columns
emotion_scores = sampled_data['emotion_scores'].apply(pd.Series)

# Join the expanded emotion scores dataframe with the original data
data_with_emotions = sampled_data.join(emotion_scores)

# Display the first few rows to verify the expansion
data_with_emotions.head()


uid_review           profile  anime_uid  \
0       27270           Elfsire        120   
1      228044      merryfistmas        227   
2       21744         ravenrage       3303   
3      248039  Tyrannicswine117      32615   
4      215510       ThePunzalan      14813   

                                              scores  Overall  Story  \
0  {'Overall': '8', 'Story': '8', 'Animation': '1...        8      8   
1  {'Overall': '10', 'Story': '10', 'Animation': ...       10     10   
2  {'Overall': '6', 'Story': '3', 'Animation': '8...        6      3   
3  {'Overall': '8', 'Story': '9', 'Animation': '8...        8      9   
4  {'Overall': '9', 'Story': '9', 'Animation': '9...        9      9   

   Animation  Sound  Character  Enjoyment  ...  \
0         10      8         10          7  ...   
1         10     10         10         10  ...   
2          8      5          6          7  ...   
3          8      8          9          8  ...   
4          9      9         10          9  ...   

                                     favorites_anime  \
0  ['14713', '14289', '4081', '14467', '16498', '...   
1  ['18679', '22135', '227', '23847', '5114', '11...   
2                                            ['134']   
3  ['5114', '4181', '9253', '18679', '1', '2001',...   
4  ['14813', '30831', '15809', '4224', '31715', '...   

                                             link_y  \
0           https://myanimelist.net/profile/Elfsire   
1      https://myanimelist.net/profile/merryfistmas   
2         https://myanimelist.net/profile/ravenrage   
3  https://myanimelist.net/profile/Tyrannicswine117   
4       https://myanimelist.net/profile/ThePunzalan   

                                      emotion_scores     anger   disgust  \
0  {'anger': 0.010413373820483685, 'disgust': 0.0...  0.010413  0.000746   
1  {'anger': 0.18671022355556488, 'disgust': 0.00...  0.186710  0.002664   
2  {'anger': 0.007550527341663837, 'disgust': 0.0...  0.007551  0.001111   
3  {'anger': 0.5558895468711853, 'disgust': 0.003...  0.555890  0.003051   
4  {'anger': 0.011294235475361347, 'disgust': 0.0...  0.011294  0.000683   

       fear       joy   neutral   sadness  surprise  
0  0.197999  0.028194  0.013339  0.689401  0.059908  
1  0.069133  0.091217  0.139144  0.045635  0.465496  
2  0.001951  0.658758  0.077451  0.017352  0.235827  
3  0.173663  0.078683  0.091056  0.014780  0.082877  
4  0.000702  0.938779  0.007481  0.031252  0.009808  

[5 rows x 34 columns]

## Baseline Classification Model

### Classification Model for Mood Categorization

### Rule-Based Labeling

In [63]:
def create_mood_label(row):
    if row['joy'] > 0.5:
        return 'Happy'
    elif row['sadness'] > 0.5:
        return 'Sad'
    elif row['anger'] > 0.5:
        return 'Angry'
    elif row['disgust'] > 0.5:
        return 'Disgust'
    elif row['fear'] > 0.5:
        return 'Fear'
    elif row['surprise'] > 0.5:
        return 'Excitement'
    else:
        return 'Neutral'

data_with_emotions['mood_label'] = data_with_emotions.apply(create_mood_label, axis=1)
data_with_emotions['mood_label'].value_counts()


mood_label
Fear          8975
Neutral       7069
Sad           5518
Happy         4686
Angry         4410
Excitement    4315
Disgust         15
Name: count, dtype: int64

In [90]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import FunctionTransformer


# Prepare the data - Features and Labels
X = data_with_emotions[['uid_anime', 'title', 'genre', 'popularity']]
y = data_with_emotions['mood_label']

numerical_feature = ['uid_anime', 'popularity']
categorical_feature = ['title', 'genre']

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)


# Categorical feature preprocessing
categorical_transformer = OneHotEncoder(handle_unknown='ignore')

# Numerical feature preprocessing with missing value handling
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean'))
])

# Combine preprocessing for different types of features
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', categorical_transformer, categorical_feature),
        ('num', numerical_transformer, numerical_feature)
    ]
)

# Apply preprocessing
# We need to apply preprocessing to train and test sets separately
X_train_transformed = preprocessor.fit_transform(X_train)
X_test_transformed = preprocessor.transform(X_test)

In [91]:
# Initialize the classifier
classifier = RandomForestClassifier(random_state=13)

# Train the classifier with transformed training data
classifier.fit(X_train_transformed, y_train)

# Perform cross-validation
# Note: For cross-validation, you need to preprocess the entire dataset
X_transformed = preprocessor.fit_transform(X)
cross_val_scores = cross_val_score(classifier, X_transformed, y, cv=5)
print(f'Cross-Validation Scores: {cross_val_scores.mean():.2f}')

# Make predictions on the transformed test set
y_pred = classifier.predict(X_test_transformed)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy on Test Data: {accuracy:.2f}')

# Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:\n', cm)


Cross-Validation Scores: 0.98
Accuracy on Test Data: 0.98
Confusion Matrix:
 [[ 840    0    0    0    1   16    0]
 [   0    3    0    0    0    0    0]
 [   0    0  831    0    0   23    0]
 [   0    0    1 1756    1   21    0]
 [   0    0    0    0  936   29    0]
 [   1    0    0    0    2 1414    0]
 [   0    0    0    0    0   25 1098]]


## Interpretation

### Model Features
- Numerical Features (uid_anime, popularity): These features likely provide information about the identity and popularity of each anime. uid_anime is a unique identifier, and popularity reflects how well-known or frequently watched an anime is.

- Categorical Features (title, genre): The title may not offer much predictive power due to its uniqueness for each anime, but genre is crucial as it categorically describes the type of content in an anime, which can significantly influence its mood.

### Cross-Validation Scores
Mean of Cross-Validation Scores: Averaging at {cross_val_scores.mean():.2f}, this metric indicates the model's average performance across different subsets of the dataset. A higher score suggests better generalization, meaning the model performs consistently across different data samples.

### Accuracy on Test Data
- Accuracy: With an accuracy of {accuracy:.2f} on the test set, this metric tells you how often the model correctly predicts the mood label. A high accuracy is desirable, indicating that the model performs well on unseen data. However, accuracy alone might not give a complete picture, especially if the dataset is imbalanced (i.e., some mood labels are much more common than others).

### Confusion Matrix
Confusion Matrix: The confusion matrix cm provides a detailed breakdown of the model's predictions. It shows the number of correct and incorrect predictions for each mood label, classified into true positives, false positives, true negatives, and false negatives. Analyzing the confusion matrix can help identify if the model is particularly good or bad at predicting specific mood labels.

### Interpretation and Considerations
- The model shows high accuracy and consistent cross-validation scores, indicating that the chosen features (especially genre and popularity) are good predictors for the mood label of an anime.

- The usefulness of uid_anime and title in predicting mood might be limited. uid_anime is more about identification, and title varies widely.

- If there are mood labels with poor prediction rates, as suggested by the confusion matrix, it could indicate a need for more representative data for those labels or a reevaluation of the features used.

- Consider the balance of your dataset. If some mood labels are underrepresented, the model's ability to predict those moods might be limited, which can be reflected in the confusion matrix.

### Conclusion
Overall, the combination of genre and popularity as features seems to be key in predicting the mood of an anime, supplemented by the unique identifiers of each anime. The model's performance as indicated by the accuracy and confusion matrix suggests how well these features work together to predict moods in your dataset.

## Tuning Classification Model Using Popularity & Genre

In [92]:
# Prepare the data - Features and Labels
X = data_with_emotions[['genre', 'popularity']]
y = data_with_emotions['mood_label']

numerical_feature = ['popularity']
categorical_feature = ['genre']

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)


# Categorical feature preprocessing
categorical_transformer = OneHotEncoder(handle_unknown='ignore')

# Numerical feature preprocessing with missing value handling
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean'))
])

# Combine preprocessing for different types of features
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', categorical_transformer, categorical_feature),
        ('num', numerical_transformer, numerical_feature)
    ]
)

# Apply preprocessing
# We need to apply preprocessing to train and test sets separately
X_train_transformed = preprocessor.fit_transform(X_train)
X_test_transformed = preprocessor.transform(X_test)

In [93]:
# Initialize the classifier
mood_classifier = RandomForestClassifier(random_state=13)

# Train the classifier with transformed training data
mood_classifier.fit(X_train_transformed, y_train)

# Perform cross-validation
# Note: For cross-validation, you need to preprocess the entire dataset
X_transformed = preprocessor.fit_transform(X)
cross_val_scores = cross_val_score(mood_classifier, X_transformed, y, cv=5)
print(f'Cross-Validation Scores: {cross_val_scores.mean():.2f}')

# Make predictions on the transformed test set
y_pred = mood_classifier.predict(X_test_transformed)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy on Test Data: {accuracy:.2f}')

# Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:\n', cm)


Cross-Validation Scores: 0.98
Accuracy on Test Data: 0.98
Confusion Matrix:
 [[ 840    0    1    0    2   12    2]
 [   0    3    0    0    0    0    0]
 [   1    0  832    0    1   20    0]
 [   1    0    2 1758    2   15    1]
 [   1    0    2    0  939   21    2]
 [   5    0    3    4    6 1398    1]
 [   0    0    0    1    1   22 1099]]


### Identifying Super-users

Using master_data

In [19]:
# Calculate the number of interactions per user
interaction_counts = master_data.groupby('profile').size()

# Using percentiles to define super-users
percentile_threshold = 95  # Top 5% of users
threshold = interaction_counts.quantile(percentile_threshold / 100)
super_users = interaction_counts[interaction_counts > threshold].index.tolist()

print(f"Threshold for super-users (top {100 - percentile_threshold}%): {threshold:.2f}")
print(f"Identified {len(super_users)} super-users.")


Threshold for super-users (top 5%): 56.00
Identified 970 super-users.


Using data_with_emotions subset - approx. 34k

In [6]:
# Calculate the number of interactions per user
interaction_counts = data_with_emotions.groupby('profile').size()

# Using percentiles to define super-users
percentile_threshold = 95  # Top 5% of users
threshold = interaction_counts.quantile(percentile_threshold / 100)
super_users = interaction_counts[interaction_counts > threshold].index.tolist()

print(f"Threshold for super-users (top {100 - percentile_threshold}%): {threshold:.2f}")
print(f"Identified {len(super_users)} super-users.")


Threshold for super-users (top 5%): 10.00
Identified 497 super-users.


### New dataset with and without super-users

In [ ]:
#super_users = 

## Baseline Model

### Neural Collaborative Filtering (NCF)
NCF models use neural networks to learn user-item interaction patterns, providing a more flexible and powerful way to model complex relationships compared to traditional matrix factorization methods.

In [22]:
pip install tensorflow

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 208.8 MB 8.1 kB/s  eta 0:00:01████████████████▍             | 119.6 MB 79.2 MB/s eta 0:00:02
     |████████████████████████████████| 394 kB 72.3 MB/s eta 0:00:01
  Using cached flatbuffers-23.5.26-py2.py3-none-any.whl (26 kB)
     |████████████████████████████████| 1.2 MB 44.7 MB/s eta 0:00:01
  Using cached tensorflow_estimator-2.15.0-py2.py3-none-any.whl (441 kB)
  Using cached termcolor-2.4.0-py3-none-any.whl (7.7 kB)
     |████████████████████████████████| 1.9 MB 39.6 MB/s eta 0:00:01
     |████████████████████████████████| 2.7 MB 80.8 MB/s eta 0:00:01
  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)
  Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)
     |████████████████████████████████| 20.6 MB 46.9 MB/s eta 0:00:01
  Using cached absl_py-2.0.0-py3-none-any.whl (130 kB)
     |████████████████████████████████| 9.6 MB 66.6 MB/s eta 0:00:01
  Using c

In [23]:
pip install keras

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [7]:
from keras.models import Model
from keras.layers import Embedding, Flatten, Input, concatenate, Dense
from keras.optimizers import Adam
from sklearn.model_selection import train_test_split
from tensorflow import keras

# Columns: 'uid_review' for users, 'anime_uid' for items, and 'Overall' for ratings
# Convert user and item IDs to categorical variables
data_with_emotions['uid_review'] = data_with_emotions['uid_review'].astype('category')
data_with_emotions['anime_uid'] = data_with_emotions['anime_uid'].astype('category')

# Map user and item IDs to integers
user_id_mapping = {id: i for i, id in enumerate(data_with_emotions['uid_review'].cat.categories)}
item_id_mapping = {id: i for i, id in enumerate(data_with_emotions['anime_uid'].cat.categories)}

# Apply mapping
data_with_emotions['user'] = data_with_emotions['uid_review'].map(user_id_mapping)
data_with_emotions['item'] = data_with_emotions['anime_uid'].map(item_id_mapping)

In [10]:
# Prepare the data for the model
X = data_with_emotions[['user', 'item']].values
y = data_with_emotions['Overall'].values

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)

# Neural Collaborative Filtering model architecture
user_input = Input(shape=(1,))
item_input = Input(shape=(1,))
user_embedding = Embedding(len(user_id_mapping), 15, input_length=1)(user_input)
item_embedding = Embedding(len(item_id_mapping), 15, input_length=1)(item_input)
user_vec = Flatten()(user_embedding)
item_vec = Flatten()(item_embedding)
concat = concatenate([user_vec, item_vec])
dense = Dense(64, activation='relu')(concat)
output = Dense(1)(dense)
model = Model([user_input, item_input], output)

# Compile and train the model
model.compile(optimizer=Adam(0.001), loss='mean_squared_error')
model.fit([X_train[:, 0], X_train[:, 1]], y_train, batch_size=64, epochs=5, validation_split=0.1)

# Evaluate the model
model.evaluate([X_test[:, 0], X_test[:, 1]], y_test)

Epoch 1/5
394/394 [==============================] - 1s 2ms/step - loss: 20.6236 - val_loss: 3.8344
Epoch 2/5
394/394 [==============================] - 1s 2ms/step - loss: 2.0858 - val_loss: 2.7484
Epoch 3/5
394/394 [==============================] - 1s 2ms/step - loss: 0.5856 - val_loss: 2.6623
Epoch 4/5
394/394 [==============================] - 1s 2ms/step - loss: 0.3072 - val_loss: 2.6076
Epoch 5/5
219/219 [==============================] - 0s 297us/step - loss: 2.5418


2.5418050289154053

## Interpretation

The model is a Neural Collaborative Filtering (NCF) model, which is designed to predict user ratings for items (anime) based on learned user and item embeddings.

### Model Architecture:

- Model has separate embedding layers for users (user_embedding) and items (item_embedding), each with an embedding size of 15. This means the model learns a 15-dimensional vector to represent each user and each item.

- Embeddings are flattened and then concatenated.

- After concatenation, the data passes through a dense layer with 64 neurons and ReLU activation, which allows the model to learn non-linear relationships.

- The final output layer has 1 neuron (since this is a regression problem where you're predicting a rating value).

### Compilation and Training:

- The model is compiled with the Adam optimizer and mean squared error (MSE) loss.

- The model is trained for 5 epochs with a batch size of 64 and uses 10% of the training data for validation.


### Model Evaluation:

- After training, the model is evaluated on the test set, resulting in a MSE of approximately 2.61.

- The MSE is a measure of the average squared difference between the actual ratings and the ratings predicted by the model. A lower MSE indicates better performance.

- This MSE value means that on average, the predicted rating deviates from the actual rating by the square root of 2.612, which is around 1.62. 


### Interpreting the Results:

- The scale of ratings is 1 to 10, an average deviation of around 1.6 might be considered moderate. It shows that the model has learned to some extent but still has room for improvement.


### Next Steps for Improvement:

- Consider training the model for more epochs or tuning other hyperparameters.
- Experiment with different architectures, like adding more dense layers or changing the number of neurons.
- Explore advanced techniques like regularization (to prevent overfitting) and learning rate scheduling.
- Consider content-based features (like genres or synopses), consider a hybrid model that combines collaborative and content-based approaches.



## Hybrid Deep Learning Models
Combine features from both content-based and collaborative filtering in a deep learning architecture, allowing the model to learn complex interactions between content and user preferences.

In [35]:
#pip install scikit-surprise

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 771 kB 3.3 MB/s eta 0:00:01
  ERROR: Command errored out with exit status 1:
   command: /Library/Developer/CommandLineTools/usr/bin/python3 -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/zg/2vpc_k5518g10mcclm90c_sw0000gn/T/pip-install-mxlyhv_9/scikit-surprise_2e6b0445d5dc46869fe3afbc2577d17f/setup.py'"'"'; __file__='"'"'/private/var/folders/zg/2vpc_k5518g10mcclm90c_sw0000gn/T/pip-install-mxlyhv_9/scikit-surprise_2e6b0445d5dc46869fe3afbc2577d17f/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /private/var/folders/zg/2vpc_k5518g10mcclm90c_sw0000gn/T/pip-wheel-4h7z3_dx
       cwd: /private/var/f

### Content-Based Filtering
Feature Extraction from Synopsis using TF-IDF

In [55]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from keras.layers import Embedding, Input, Flatten, concatenate, Dense
from keras.models import Model
from keras.optimizers import Adam

# Fill NaN values in 'synopsis' with empty string
data_with_emotions['synopsis'].fillna('', inplace=True)

# Split data
train_data, test_data = train_test_split(data_with_emotions, test_size=0.2, random_state=13)

# Apply TF-IDF to the training set and transform the test set
tfidf = TfidfVectorizer(stop_words='english')
train_content_features = tfidf.fit_transform(train_data['synopsis'])
test_content_features = tfidf.transform(test_data['synopsis'])

# Convert sparse TF-IDF matrix to a dense format for both train and test sets
train_content_features_dense = train_content_features.todense()
test_content_features_dense = test_content_features.todense()


### Prepare Collaborative Filtering Data
Prepare user-item interaction data and create embeddings for collaborative filtering

In [56]:
# Prepare user and item interaction data for both train and test sets
train_user_ids = train_data['uid_review'].astype('category').cat.codes.values
train_item_ids = train_data['anime_uid'].astype('category').cat.codes.values
test_user_ids = test_data['uid_review'].astype('category').cat.codes.values
test_item_ids = test_data['anime_uid'].astype('category').cat.codes.values

num_users = data_with_emotions['uid_review'].nunique()
num_items = data_with_emotions['anime_uid'].nunique()
embedding_size = 20

# Neural Collaborative Filtering model architecture
user_input = Input(shape=(1,))
item_input = Input(shape=(1,))
content_input = Input(shape=(train_content_features_dense.shape[1],))

user_embedding = Embedding(num_users, embedding_size, input_length=1)(user_input)
item_embedding = Embedding(num_items, embedding_size, input_length=1)(item_input)
user_vec = Flatten()(user_embedding)
item_vec = Flatten()(item_embedding)


### Hybrid Model Architecture
- Combine content-based and collaborative filtering features in a neural network

- Evaluate model with MAE

In [57]:
from keras.metrics import MeanAbsoluteError

concatenated = concatenate([user_vec, item_vec, content_input])
dense_layer_1 = Dense(128, activation='relu')(concatenated)
output = Dense(1, activation='linear')(dense_layer_1)

model_hybrid = Model(inputs=[user_input, item_input, content_input], outputs=output)
model_hybrid.compile(optimizer=Adam(0.001), loss='mean_squared_error')

# Set a smaller batch size
batch_size = 32

# Train the Model
model_hybrid.fit(
    [train_user_ids, train_item_ids, train_content_features_dense],
    train_data['Overall'].values, 
    batch_size=batch_size, 
    epochs=5, 
    validation_split=0.1
)

# Evaluate the Model using MSE
mse_loss = model_hybrid.evaluate(
    [test_user_ids, test_item_ids, test_content_features_dense], 
    test_data['Overall'].values
)
print(f"Mean Squared Error on Test Data: {mse_loss}")

# Predict on the test data
test_predictions = model_hybrid.predict(
    [test_user_ids, test_item_ids, test_content_features_dense]
)

# Calculate MAE
mae_metric = MeanAbsoluteError()
mae_metric.update_state(test_data['Overall'].values, test_predictions)
mae = mae_metric.result().numpy()
print(f"Mean Absolute Error on Test Data: {mae}")


Epoch 1/5
788/788 [==============================] - 11s 14ms/step - loss: 8.1819 - val_loss: 2.9647
Epoch 2/5
788/788 [==============================] - 20s 25ms/step - loss: 1.4987 - val_loss: 2.4187
Epoch 3/5
788/788 [==============================] - 42s 54ms/step - loss: 0.6136 - val_loss: 2.2877
Epoch 4/5
788/788 [==============================] - 25s 32ms/step - loss: 0.3200 - val_loss: 2.2362
Epoch 5/5
219/219 [==============================] - 2s 7ms/step - loss: 2.2556
Mean Squared Error on Test Data: 2.255605697631836
219/219 [==============================] - 1s 5ms/step
Mean Absolute Error on Test Data: 1.0276234149932861


## Interpretation

### Training Process (Epochs 1-5)
- Loss Over Epochs: The model's loss on the training set significantly decreases from 7.7444 in the first epoch to 0.2442 by the fifth epoch. This indicates that the model is learning and improving its predictions on the training data over each epoch.

- Validation Loss: The validation loss also decreases over epochs, from 2.9592 to 2.2254, which is a positive sign. It suggests that the model is not just memorizing the training data but is also improving its performance on unseen data (validation set).

### Final Evaluation
- Test Loss: The model achieves a loss of 2.2711 on the test set. This is the mean squared error (MSE) between the predicted ratings and the actual ratings in the test data.


### Interpretation
- Model Learning: The decreasing trend in loss over epochs indicates that the model is effectively learning from the training data. However, the initial high loss suggests that the model may have started with predictions far from the actual values.

- Overfitting Check: There is no significant divergence between training loss and validation loss, which is a good sign. A large gap would have indicated overfitting, where the model performs well on training data but poorly on unseen data.

- Mean Squared Error: An MSE of 2.2711 on a rating scale (1-10) indicates that the model's predictions are, on average, about 1.51 units away from the actual ratings (since RMSE = sqrt(MSE) ≈ 1.51). The acceptability of this error depends on the context of your application. For a 1-10 scale, this deviation is moderate.

### Overview

In summary, the model is learning and improving its prediction accuracy over time, as evidenced by decreasing training and validation loss. The final MSE on the test set suggests moderate predictive accuracy, with potential areas for improvement.

## Tuning the Hybrid Model

### Adjust the Model Architecture
Modifying the architecture of the neural network by adding more layers, changing the number of neurons, and adding dropout layers to prevent overfitting.

In [17]:
from keras.layers import Dropout

# Adjusted Neural Collaborative Filtering model architecture
user_input = Input(shape=(1,))
item_input = Input(shape=(1,))
content_input = Input(shape=(train_content_features_dense.shape[1],))

user_embedding = Embedding(num_users, embedding_size, input_length=1)(user_input)
item_embedding = Embedding(num_items, embedding_size, input_length=1)(item_input)
user_vec = Flatten()(user_embedding)
item_vec = Flatten()(item_embedding)

concatenated = concatenate([user_vec, item_vec, content_input])
dense_layer_1 = Dense(256, activation='relu')(concatenated)
dropout_1 = Dropout(0.3)(dense_layer_1)
dense_layer_2 = Dense(128, activation='relu')(dropout_1)
output = Dense(1, activation='linear')(dense_layer_2)

model_hybrid = Model(inputs=[user_input, item_input, content_input], outputs=output)


### Reducing the Learning Rate

In [20]:
# Adjust the learning rate
model_hybrid.compile(optimizer=Adam(0.0005), loss='mean_squared_error')


### Train Adjusted Hybrid Model
Increase epochs from 5 to 10. 

In [21]:
model_hybrid.fit(
    [train_user_ids, train_item_ids, train_content_features_dense],
    train_data['Overall'].values, 
    batch_size=32,
    epochs=10,
    validation_split=0.1
)

Epoch 1/10
788/788 [==============================] - 22s 28ms/step - loss: 6.7640 - val_loss: 3.3927
Epoch 2/10
788/788 [==============================] - 23s 30ms/step - loss: 2.1120 - val_loss: 2.6839
Epoch 3/10
788/788 [==============================] - 28s 36ms/step - loss: 0.9957 - val_loss: 2.4714
Epoch 4/10
788/788 [==============================] - 47s 59ms/step - loss: 0.6179 - val_loss: 2.4497
Epoch 5/10
788/788 [==============================] - 40s 50ms/step - loss: 0.4822 - val_loss: 2.3505
Epoch 6/10
788/788 [==============================] - 37s 47ms/step - loss: 0.4030 - val_loss: 2.2939
Epoch 7/10
788/788 [==============================] - 37s 47ms/step - loss: 0.3390 - val_loss: 2.1830
Epoch 8/10
788/788 [==============================] - 36s 46ms/step - loss: 0.2811 - val_loss: 2.1603
Epoch 9/10
788/788 [==============================] - 35s 45ms/step - loss: 0.2436 - val_loss: 2.0943
Epoch 10/10
788/788 [==============================] - 34s 44ms/step - loss: 0.216

### Evaluate Adjusted Hybrid Model with Mean Absolute Error (MAE)

In [23]:
from sklearn.metrics import mean_absolute_error

# Predict on the test set
test_predictions = model_hybrid.predict(
    [test_user_ids, test_item_ids, test_content_features_dense]
)

# Calculate MAE
mae = mean_absolute_error(test_data['Overall'].values, test_predictions)
mae


219/219 [==============================] - 1s 3ms/step


0.9768970113680955

## Interpretation

## Feature Engineering
Converting 'Birthday' to Age

In [25]:
from datetime import datetime

# Convert 'birthday' to datetime and then to age
current_year = datetime.now().year
data_with_emotions['age'] = data_with_emotions['birthday'].apply(lambda x: current_year - pd.to_datetime(x, errors='coerce').year)

# Sanity check
data_with_emotions.head()

uid_review           profile anime_uid  \
0      27270           Elfsire       120   
1     228044      merryfistmas       227   
2      21744         ravenrage      3303   
3     248039  Tyrannicswine117     32615   
4     215510       ThePunzalan     14813   

                                              scores  Overall  Story  \
0  {'Overall': '8', 'Story': '8', 'Animation': '1...        8      8   
1  {'Overall': '10', 'Story': '10', 'Animation': ...       10     10   
2  {'Overall': '6', 'Story': '3', 'Animation': '8...        6      3   
3  {'Overall': '8', 'Story': '9', 'Animation': '8...        8      9   
4  {'Overall': '9', 'Story': '9', 'Animation': '9...        9      9   

   Animation  Sound  Character  Enjoyment  ...     anger   disgust      fear  \
0         10      8         10          7  ...  0.010413  0.000746  0.197999   
1         10     10         10         10  ...  0.186710  0.002664  0.069133   
2          8      5          6          7  ...  0.007551  0.001111  0.001951   
3          8      8          9          8  ...  0.555890  0.003051  0.173663   
4          9      9         10          9  ...  0.011294  0.000683  0.000702   

        joy   neutral   sadness  surprise   user  item   age  
0  0.028194  0.013339  0.689401  0.059908   2549    62  32.0  
1  0.091217  0.139144  0.045635  0.465496  14674   111  30.0  
2  0.658758  0.077451  0.017352  0.235827   2087   887  33.0  
3  0.078683  0.091056  0.014780  0.082877  16645  2249  26.0  
4  0.938779  0.007481  0.031252  0.009808  13400  1652   NaN  

[5 rows x 37 columns]

## Tuning Hybrid Model with Feature Engineering

- TF-IDF: Adjusted with parameters max_features and ngram_range.
- Regularization: Added L1 and L2 regularization to the first dense layer.
- Cross-Validation: Implemented with 5-fold cross

### Refine TF-IDF Features

In [30]:
from keras.layers import Embedding, Input, Flatten, concatenate, Dense, Dropout
from keras.optimizers import Adam
from keras.regularizers import l1_l2
from keras import backend as K
from sklearn.model_selection import train_test_split, KFold

# Fill NaN values in 'synopsis' with empty string
data_with_emotions['synopsis'].fillna('', inplace=True)

# Split data
train_data, test_data = train_test_split(data_with_emotions, test_size=0.2, random_state=13)

# Apply TF-IDF to the training set and transform the test set
tfidf = TfidfVectorizer(stop_words='english', max_features=1000, ngram_range=(1, 2))
train_content_features = tfidf.fit_transform(train_data['synopsis'])
test_content_features = tfidf.transform(test_data['synopsis'])

# Convert sparse TF-IDF matrix to a dense format for both train and test sets
train_content_features_dense = train_content_features.todense()
test_content_features_dense = test_content_features.todense()

In [31]:
# Prepare user and item interaction data for both train and test sets
train_user_ids = train_data['uid_review'].astype('category').cat.codes.values
train_item_ids = train_data['anime_uid'].astype('category').cat.codes.values
test_user_ids = test_data['uid_review'].astype('category').cat.codes.values
test_item_ids = test_data['anime_uid'].astype('category').cat.codes.values

num_users = data_with_emotions['uid_review'].nunique()
num_items = data_with_emotions['anime_uid'].nunique()
embedding_size = 20


In [32]:
# Define a function to create the model
def create_model():
    user_input = Input(shape=(1,))
    item_input = Input(shape=(1,))
    content_input = Input(shape=(train_content_features_dense.shape[1],))

    user_embedding = Embedding(num_users, embedding_size, input_length=1)(user_input)
    item_embedding = Embedding(num_items, embedding_size, input_length=1)(item_input)
    user_vec = Flatten()(user_embedding)
    item_vec = Flatten()(item_embedding)

    concatenated = concatenate([user_vec, item_vec, content_input])
    dense_layer_1 = Dense(256, activation='relu', kernel_regularizer=l1_l2(l1=0.01, l2=0.01))(concatenated)
    dropout_1 = Dropout(0.3)(dense_layer_1)
    dense_layer_2 = Dense(128, activation='relu')(dropout_1)
    output = Dense(1, activation='linear')(dense_layer_2)

    model = Model(inputs=[user_input, item_input, content_input], outputs=output)
    model.compile(optimizer=Adam(0.0005), loss='mean_squared_error')
    return model

In [34]:
from sklearn.model_selection import KFold

# Define K-Fold Cross Validator
kfold = KFold(n_splits=5, shuffle=True, random_state=13)

# Convert features and labels to numpy arrays for indexing
train_user_ids = np.array(train_user_ids)
train_item_ids = np.array(train_item_ids)
train_content_features_dense = np.array(train_content_features_dense)
train_labels = np.array(train_data['Overall'].values)

# K-fold Cross Validation model evaluation
fold_no = 1
for train_indices, test_indices in kfold.split(train_user_ids):
    # Split data
    X_train_user, X_test_user = train_user_ids[train_indices], train_user_ids[test_indices]
    X_train_item, X_test_item = train_item_ids[train_indices], train_item_ids[test_indices]
    X_train_content, X_test_content = train_content_features_dense[train_indices], train_content_features_dense[test_indices]
    y_train, y_test = train_labels[train_indices], train_labels[test_indices]

    # Create a new model instance
    model = create_model()
    
    # Fit data to model
    model.fit(
        [X_train_user, X_train_item, X_train_content],
        y_train, 
        batch_size=32, 
        epochs=10, 
        validation_split=0.1
    )
    
    # Evaluate the model
    scores = model.evaluate(
        [X_test_user, X_test_item, X_test_content],
        y_test
    )
    print(f'Score for fold {fold_no}: {scores}')
    fold_no += 1

Epoch 1/10
630/630 [==============================] - 4s 5ms/step - loss: 18.9592 - val_loss: 4.7097
Epoch 2/10
630/630 [==============================] - 3s 5ms/step - loss: 4.4315 - val_loss: 4.2943
Epoch 3/10
630/630 [==============================] - 3s 5ms/step - loss: 2.8398 - val_loss: 3.8752
Epoch 4/10
630/630 [==============================] - 2s 4ms/step - loss: 1.8742 - val_loss: 3.7399
Epoch 5/10
630/630 [==============================] - 3s 4ms/step - loss: 1.4446 - val_loss: 3.2697
Epoch 6/10
630/630 [==============================] - 3s 4ms/step - loss: 1.2434 - val_loss: 3.2069
Epoch 7/10
630/630 [==============================] - 3s 4ms/step - loss: 1.1451 - val_loss: 3.2018
Epoch 8/10
630/630 [==============================] - 3s 5ms/step - loss: 1.0839 - val_loss: 3.0934
Epoch 9/10
630/630 [==============================] - 3s 5ms/step - loss: 1.0343 - val_loss: 3.0077
Epoch 10/10
175/175 [==============================] - 0s 973us/step - loss: 3.1458
Score for fold 

Epoch 1/10
630/630 [==============================] - 4s 5ms/step - loss: 19.1326 - val_loss: 4.7790
Epoch 2/10
630/630 [==============================] - 4s 6ms/step - loss: 4.3556 - val_loss: 4.3834
Epoch 3/10
630/630 [==============================] - 4s 6ms/step - loss: 2.7473 - val_loss: 3.9054
Epoch 4/10
630/630 [==============================] - 3s 5ms/step - loss: 1.7810 - val_loss: 3.5295
Epoch 5/10
630/630 [==============================] - 4s 6ms/step - loss: 1.3877 - val_loss: 3.4122
Epoch 6/10
630/630 [==============================] - 5s 7ms/step - loss: 1.1937 - val_loss: 3.2323
Epoch 7/10
630/630 [==============================] - 4s 7ms/step - loss: 1.0974 - val_loss: 3.1743
Epoch 8/10
630/630 [==============================] - 5s 7ms/step - loss: 1.0576 - val_loss: 3.0950
Epoch 9/10
630/630 [==============================] - 4s 6ms/step - loss: 1.0145 - val_loss: 3.0844
Epoch 10/10
175/175 [==============================] - 0s 1ms/step - loss: 3.1871
Score for fold 2:

Epoch 1/10
630/630 [==============================] - 4s 6ms/step - loss: 19.0967 - val_loss: 4.6045
Epoch 2/10
630/630 [==============================] - 4s 6ms/step - loss: 4.4493 - val_loss: 4.3827
Epoch 3/10
630/630 [==============================] - 4s 6ms/step - loss: 2.9274 - val_loss: 4.1572
Epoch 4/10
630/630 [==============================] - 4s 6ms/step - loss: 1.8982 - val_loss: 3.7194
Epoch 5/10
630/630 [==============================] - 4s 7ms/step - loss: 1.4001 - val_loss: 3.4397
Epoch 6/10
630/630 [==============================] - 4s 6ms/step - loss: 1.1982 - val_loss: 3.4108
Epoch 7/10
630/630 [==============================] - 4s 6ms/step - loss: 1.1064 - val_loss: 3.2349
Epoch 8/10
630/630 [==============================] - 4s 6ms/step - loss: 1.0499 - val_loss: 3.2202
Epoch 9/10
630/630 [==============================] - 3s 6ms/step - loss: 1.0187 - val_loss: 3.3050
Epoch 10/10
175/175 [==============================] - 0s 815us/step - loss: 3.0832
Score for fold 

Epoch 1/10
630/630 [==============================] - 4s 6ms/step - loss: 19.0609 - val_loss: 4.7220
Epoch 2/10
630/630 [==============================] - 4s 6ms/step - loss: 4.4271 - val_loss: 4.3346
Epoch 3/10
630/630 [==============================] - 4s 6ms/step - loss: 2.7649 - val_loss: 4.0568
Epoch 4/10
630/630 [==============================] - 4s 6ms/step - loss: 1.7686 - val_loss: 3.4711
Epoch 5/10
630/630 [==============================] - 4s 6ms/step - loss: 1.3472 - val_loss: 3.3818
Epoch 6/10
630/630 [==============================] - 4s 6ms/step - loss: 1.1764 - val_loss: 3.2507
Epoch 7/10
630/630 [==============================] - 4s 6ms/step - loss: 1.0904 - val_loss: 3.2084
Epoch 8/10
630/630 [==============================] - 4s 6ms/step - loss: 1.0558 - val_loss: 3.0732
Epoch 9/10
630/630 [==============================] - 4s 6ms/step - loss: 1.0110 - val_loss: 3.1570
Epoch 10/10
175/175 [==============================] - 0s 1ms/step - loss: 3.1210
Score for fold 4:

Epoch 1/10
630/630 [==============================] - 4s 6ms/step - loss: 19.3148 - val_loss: 4.7460
Epoch 2/10
630/630 [==============================] - 4s 6ms/step - loss: 4.4447 - val_loss: 4.4281
Epoch 3/10
630/630 [==============================] - 4s 6ms/step - loss: 2.9425 - val_loss: 3.8823
Epoch 4/10
630/630 [==============================] - 3s 5ms/step - loss: 1.8708 - val_loss: 3.5950
Epoch 5/10
630/630 [==============================] - 3s 6ms/step - loss: 1.4131 - val_loss: 3.3857
Epoch 6/10
630/630 [==============================] - 4s 7ms/step - loss: 1.2047 - val_loss: 3.2010
Epoch 7/10
630/630 [==============================] - 4s 6ms/step - loss: 1.0981 - val_loss: 3.2211
Epoch 8/10
630/630 [==============================] - 3s 5ms/step - loss: 1.0452 - val_loss: 3.1745
Epoch 9/10
630/630 [==============================] - 4s 7ms/step - loss: 1.0188 - val_loss: 3.2555
Epoch 10/10
175/175 [==============================] - 0s 1ms/step - loss: 3.0487
Score for fold 5:

## Interpretation

### Training Loss and Validation Loss Over Epochs:
- Decreasing Loss: For each fold, both the training loss and validation loss decrease over epochs, indicating that the model is learning and improving its predictions as training progresses.

- Training Loss: Starts high in the first epoch (around 19) and drops significantly by the 10th epoch (around 1), showing that the model is effectively learning from the training data.

- Validation Loss: Although it decreases, the validation loss doesn't decrease as sharply as the training loss. This is expected as the validation set is used to gauge the model's performance on unseen data.


### Model Evaluation on Test Data:
- Fold-wise Scores: The final loss scores on the test sets for each fold hover around 3. This suggests a consistent performance across different subsets of your data, which is a good sign of model stability.


### Interpretation:
- General Learning Trend: The model's ability to reduce loss over epochs indicates successful learning. However, the fact that the training loss is significantly lower than the validation loss could hint at overfitting, where the model performs well on training data but slightly worse on unseen data.

- Model Consistency: The similar loss scores across different folds in cross-validation indicate that the model's performance is consistent regardless of the specific subset of data it's trained on.

- Loss Magnitude: A loss of around 3, considering it's mean squared error, suggests the predictions are, on average, somewhat off from the actual values. The square root of this value gives an RMSE (Root Mean Squared Error) of approximately 1.73, which can be interpreted as the average deviation from the actual values in your rating prediction task. On a scale of 1-10, this deviation is moderate.

In [35]:
data_with_emotions['gender'].value_counts()

gender
Male          26017
Female         8489
Non-Binary      482
Name: count, dtype: int64

In [36]:
# Converting 'age' into age brackets
def age_bracket(age):
    if age < 18:
        return 'Under 18'
    elif 18 <= age < 25:
        return '18-24'
    elif 25 <= age < 35:
        return '25-34'
    elif 35 <= age < 45:
        return '35-44'
    elif 45 <= age < 55:
        return '45-54'
    else:
        return '55+'

data_with_emotions['age_bracket'] = data_with_emotions['age'].apply(age_bracket)

In [37]:
# Fill NaN values in 'synopsis' with empty string
data_with_emotions['synopsis'].fillna('', inplace=True)

## Tuning Hybrid with Gender/Age Brackets
Regression model predicts how a user might rate an anime

This code is a comprehensive process for training a neural network model on a dataset with text and categorical features. Let's break down each part of the code and then interpret the model's results:

Code Explanation
Text Preprocessing Pipeline:

A pipeline is defined using TfidfVectorizer to transform the 'synopsis' text data into a TF-IDF matrix. It removes English stopwords, limits the number of features to 1000, and considers unigrams and bigrams.
Overall Preprocessor:

ColumnTransformer is used to apply different transformations to different columns: TF-IDF vectorization to 'synopsis', one-hot encoding to 'age_bracket', and 'gender' columns.
Data Splitting:

The dataset is split into training and test sets (80-20 split). The training set is further split into a smaller training set and a validation set (90-10 split).
Data Transformation:

The preprocessor is fitted on the training data and used to transform the training, validation, and test sets. The transformed data is converted to dense format.
ID Preparation:

User and item IDs are extracted from the training, validation, and test datasets. These are categorical variables converted to numerical codes.
Model Definition:

A function create_model is defined to create a neural network model. The model uses embedding layers for user and item IDs, concatenates these with the additional preprocessed features, and passes them through dense layers.
Model Creation and Compilation:

The model is created with the appropriate input shape and compiled with the Adam optimizer and mean squared error loss function.
Model Training:

The model is trained on the training data, with the validation data used to monitor performance during training.
Model Evaluation:

The model's performance is evaluated on the test set using Mean Absolute Error (MAE) as the metric.
Optimizer Warning:

A warning indicates that the Adam optimizer version being used may run slowly on M1/M2 Macs.

In [82]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from keras.models import Model
from keras.layers import Embedding, Input, Flatten, concatenate, Dense
from keras.optimizers import Adam
from keras.regularizers import l2
from sklearn.metrics import mean_absolute_error
from sklearn.pipeline import Pipeline


# Define text preprocessing pipeline
text_pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(stop_words='english', max_features=1000, ngram_range=(1, 2)))
])

# Define overall preprocessor
preprocessor = ColumnTransformer(
    transformers=[
        ('text', text_pipeline, 'synopsis'),
        ('onehot_age', OneHotEncoder(), ['age_bracket']),
        ('onehot_gender', OneHotEncoder(), ['gender'])
    ]
)

# Split data into training and test sets
train_data, test_data = train_test_split(data_with_emotions, test_size=0.2, random_state=13)

# Further split training data into training and 10% validation set
train_data, val_data = train_test_split(train_data, test_size=0.1, random_state=13)

# Apply preprocessing
train_transformed = preprocessor.fit_transform(train_data).todense()
val_transformed = preprocessor.transform(val_data).todense()
test_transformed = preprocessor.transform(test_data).todense()

# Prepare user and item IDs for training, validation, and testing
train_user_ids = train_data['uid_review'].astype('category').cat.codes.values
train_item_ids = train_data['anime_uid'].astype('category').cat.codes.values

val_user_ids = val_data['uid_review'].astype('category').cat.codes.values
val_item_ids = val_data['anime_uid'].astype('category').cat.codes.values

test_user_ids = test_data['uid_review'].astype('category').cat.codes.values
test_item_ids = test_data['anime_uid'].astype('category').cat.codes.values


In [85]:
# Define the model
def create_model(num_users, num_items, input_shape):
    user_input = Input(shape=(1,))
    item_input = Input(shape=(1,))
    additional_input = Input(shape=(input_shape,))

    user_embedding = Embedding(num_users, 20, input_length=1)(user_input)
    item_embedding = Embedding(num_items, 20, input_length=1)(item_input)
    user_vec = Flatten()(user_embedding)
    item_vec = Flatten()(item_embedding)

    concatenated = concatenate([user_vec, item_vec, additional_input])
    dense_layer = Dense(64, activation='relu', kernel_regularizer=l2(0.01))(concatenated)
    output = Dense(1, activation='linear')(dense_layer)

    model = Model(inputs=[user_input, item_input, additional_input], outputs=output)
    model.compile(optimizer=Adam(0.0005), loss='mean_squared_error')
    return model

In [86]:
# Create the model
input_shape = train_transformed.shape[1]
rating_predictor = create_model(len(train_user_ids), len(train_item_ids), input_shape)

# Prepare inputs for the model
train_inputs = [train_user_ids, train_item_ids, train_transformed]
val_inputs = [val_user_ids, val_item_ids, val_transformed]
test_inputs = [test_user_ids, test_item_ids, test_transformed]

# Train the model
rating_predictor.fit(
    train_inputs,
    train_data['Overall'].values,
    batch_size=32,
    epochs=10,
    validation_data=(val_inputs, val_data['Overall'].values)
)

# Predict and evaluate using MAE
predictions = rating_predictor.predict(test_inputs)
mae = mean_absolute_error(test_data['Overall'].values, predictions)
print(f'Mean Absolute Error: {mae}')

Epoch 1/10
788/788 [==============================] - 6s 7ms/step - loss: 11.7429 - val_loss: 4.2300
Epoch 2/10
788/788 [==============================] - 5s 7ms/step - loss: 2.8092 - val_loss: 3.0669
Epoch 3/10
788/788 [==============================] - 4s 6ms/step - loss: 1.2951 - val_loss: 2.6411
Epoch 4/10
788/788 [==============================] - 4s 5ms/step - loss: 0.6979 - val_loss: 2.5108
Epoch 5/10
788/788 [==============================] - 4s 5ms/step - loss: 0.4569 - val_loss: 2.4548
Epoch 6/10
788/788 [==============================] - 5s 6ms/step - loss: 0.3332 - val_loss: 2.3901
Epoch 7/10
788/788 [==============================] - 5s 7ms/step - loss: 0.2598 - val_loss: 2.3425
Epoch 8/10
788/788 [==============================] - 5s 7ms/step - loss: 0.2102 - val_loss: 2.2923
Epoch 9/10
788/788 [==============================] - 5s 6ms/step - loss: 0.1739 - val_loss: 2.2789
Epoch 10/10
219/219 [==============================] - 0s 633us/step
Mean Absolute Error: 1.0291059

## Model Results Interpretation


### Training and Validation Loss:

The training loss decreases from 14.2328 to 0.1535 over 10 epochs, indicating that the model is learning effectively from the training data.
The validation loss also decreases, from 4.4708 to 2.2695, suggesting that the model generalizes well to unseen data.


### Evaluation - Mean Absolute Error (MAE)

The MAE on the test set is 1.0149. This metric indicates the average absolute difference between the predicted and actual values. Given the scale of the 'Overall' ratings (1-10), an MAE of around 1 suggests that the model's predictions are, on average, about one rating point away from the actual values.


### Features

- The model integrates textual features from the 'synopsis' (processed through TF-IDF) and categorical features from 'age_bracket' and 'gender' (one-hot encoded).

- The text features capture the semantic content of the synopses, while the categorical features provide demographic context.


### Effectiveness

The combination of these features, along with user and item IDs, gives the model a nuanced view of the interactions, likely contributing to its predictive performance.

### Conclusion
The model demonstrates effective learning and good generalization, as evidenced by decreasing training and validation loss. The MAE of 1.0149 indicates reasonable accuracy, considering the complexity of predicting precise rating values. The integration of text and categorical features, along with user and item embeddings, appears to provide a robust representation for the prediction task.

## Next Steps

### Combine Models for Playlist Generation

User Input for Mood:
- Create an interface (e.g., a web form) where users can input their mood, age, and gender.

Mood-Based Anime Filtering:
- Use the mood classification model to filter anime that matches the selected mood.

Rating Predictions:
- Use the rating prediction model to estimate how the user would rate these mood-matched anime.

Generate and Sort Playlist:
- Sort these anime based on the predicted ratings to create a playlist.

### Combining Recommendations Models
Combining collaborative and content-based filtering


In [ ]:
def hybrid_recommendation(user_id, content_rec, collaborative_rec, mood=None):
    # Hybrid model combining content and collaborative filtering
    hybrid_rec = content_rec * 0.5 + collaborative_rec * 0.5

    # If mood is specified, filter or re-rank recommendations
    if mood:
        hybrid_rec = filter_or_rank_by_mood(hybrid_rec, mood)

    return hybrid_rec


In [ ]:
def hybrid_recommendation(user_id, mood=None):
    # Get content-based recommendations
    content_based_rec = content_based_model(user_id)

    # Get collaborative filtering recommendations
    collaborative_rec = collaborative_model(user_id)

    # Get mood-based recommendations if a mood is specified
    if mood:
        mood_based_rec = mood_based_model(user_id, mood)
    else:
        mood_based_rec = {}

    # Combine recommendations
    # This is a simple average. Consider using more complex methods
    combined_rec = average(content_based_rec, collaborative_rec, mood_based_rec)

    return combined_rec


### Personalization and Playlist Generation

In [ ]:
def generate_playlist(user_id, mood=None):
    content_rec = get_content_based_recommendations(user_id)
    collaborative_rec = get_collaborative_recommendations(user_id)
    
    # Hybrid recommendations
    recommendations = hybrid_recommendation(user_id, content_rec, collaborative_rec, mood)

    # Generate playlist based on recommendations
    playlist = create_playlist(recommendations)
    return playlist


## User-interface Application

In [81]:
pip install joblib

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [94]:
from joblib import dump

# Save models
dump(mood_classifier, 'mood_classifier.joblib')
dump(rating_predictor, 'rating_predictor.joblib')


['rating_predictor.joblib']

### Create function for mood classifier pre-processing

In [96]:
def preprocess_for_mood_classifier(X, categorical_feature, numerical_feature):
    # Define the categorical and numerical transformers
    categorical_transformer = OneHotEncoder(handle_unknown='ignore')
    numerical_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='mean'))
    ])

    # Combine preprocessing for different types of features
    preprocessor = ColumnTransformer(
        transformers=[
            ('cat', categorical_transformer, categorical_feature),
            ('num', numerical_transformer, numerical_feature)
        ]
    )

    # Apply the preprocessing to the input data
    X_transformed = preprocessor.transform(X)

    return X_transformed


#### Create function for ratings predictions pre-processing

In [97]:
from sklearn.preprocessing import StandardScaler

def preprocess_for_rating_predictor(data, text_pipeline, preprocessor):
    # Extract user and item IDs
    user_ids = data['uid_review'].astype('category').cat.codes.values
    item_ids = data['anime_uid'].astype('category').cat.codes.values

    # Apply text preprocessing
    text_features = data['synopsis']
    text_transformed = text_pipeline.transform(text_features).todense()

    # Apply one-hot encoding and concatenate with text features
    categorical_features = data[['age_bracket', 'gender']]
    categorical_transformed = preprocessor.transform(categorical_features).todense()

    # Combine all features
    all_features = np.concatenate([text_transformed, categorical_transformed], axis=1)

    return user_ids, item_ids, all_features


### Create a function to filter anime by predicted mood

In [98]:
def filter_anime_by_mood(predicted_mood):
    # Filter the dataset for anime that matches the predicted mood
    return data_with_emotions[data_with_emotions['mood_label'] == predicted_mood]


### Create a function to sort anime based on predicted ratings

In [99]:
def sort_anime_by_ratings(filtered_anime, predicted_ratings):
    # Add the predicted ratings to the filtered anime dataset
    filtered_anime['predicted_rating'] = predicted_ratings
    
    # Sort the anime by predicted ratings in descending order
    sorted_anime = filtered_anime.sort_values(by='predicted_rating', ascending=False)
    
    # Return the sorted anime
    return sorted_anime


In [101]:
pip install Flask

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 99 kB 4.1 MB/s eta 0:00:011
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [104]:
from flask import Flask, request, render_template
from joblib import load

app = Flask(__name__)

# Load the trained models
mood_classifier = load('mood_classifier.joblib')
rating_predictor = load('rating_predictor.joblib')


@app.route('/', methods=['GET', 'POST'])
def index():
    if request.method == 'POST':
        # Process user input
        user_mood = request.form.get('mood')
        user_age = request.form.get('age')
        # Convert age to a numeric value
        try:
            user_age = int(user_age)
        except ValueError:
            # Handle cases where age is not a number
            user_age = train_data['age'].median()
                
        user_gender = request.form.get('gender')
        # Validate user_gender input
        if user_gender not in ['Male', 'Female', 'Non-Binary']:
            user_gender = train_data['gender'].median()
        
        # Generate playlist
        playlist = generate_playlist(user_mood, user_age, user_gender)

        return render_template('playlist.html', playlist=playlist)

    return render_template('index.html')

def generate_playlist(mood, age, gender):
    # Preprocess input for the mood classification model
    mood_input = preprocess_for_mood_classifier(mood, age, gender)
    predicted_mood = mood_classifier.predict(mood_input)

    # Filter anime based on predicted mood
    filtered_anime = filter_anime_by_mood(predicted_mood)

    # Preprocess the filtered anime for the rating prediction model
    rating_input = preprocess_for_rating_predictor(filtered_anime, age, gender)
    predicted_ratings = rating_predictor.predict(rating_input)

    # Combine predictions with anime details and sort by ratings
    sorted_playlist = sort_anime_by_ratings(predicted_ratings)

    # Return the sorted playlist as a list of recommended anime titles
    return sorted_playlist

if __name__ == '__main__':
    app.run(debug=True)


 * Serving Flask app '__main__'
 * Debug mode: on


INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5001
INFO:werkzeug:Press CTRL+C to quit
INFO:werkzeug: * Restarting with stat
Traceback (most recent call last):
  File "/Users/kariprimiano/Library/Python/3.9/lib/python/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/Users/kariprimiano/Library/Python/3.9/lib/python/site-packages/traitlets/config/application.py", line 1076, in launch_instance
    app.initialize(argv)
  File "/Users/kariprimiano/Library/Python/3.9/lib/python/site-packages/traitlets/config/application.py", line 118, in inner
    return method(app, *args, **kwargs)
  File "/Users/kariprimiano/Library/Python/3.9/lib/python/site-packages/ipykernel/kernelapp.py", line 692, in initialize
    self.init_sockets()
  File "/Users/kariprimiano/Library/Python/3.9/lib/python/site-packages/ipykernel/kernelapp.py", line 331

SystemExit: 1

## Future Considerations

- img_url, link_x: If you have the capability to process image data, the anime's poster or image could be used for mood prediction using advanced techniques like Convolutional Neural Networks (CNNs). However, this is more complex and may require substantial computational resources.